### PR TITLE
[#68] Address to recent FA2 changes (part 2)

### LINKED ISSUE #68: Account for new FA2 updates
# Clarification and motivation

There are new MRs updating FA2:
* https://gitlab.com/tzip/tzip/-/merge_requests/66
* https://gitlab.com/tzip/tzip/-/merge_requests/67

Once they are merged, we should update our code accordingly.

# Acceptance criteria

Once the above MRs are both merged our specification, contract code and tests are updated accordingly.

### PR DESCRIPTION
**Pull request by awkure**
## Description

This PR updates stablecoin contract according to recent FA2 changes, specifically [!66](https://gitlab.com/tzip/tzip/-/merge_requests/66) and [!67](https://gitlab.com/tzip/tzip/-/merge_requests/67)

But note, that [!67](https://gitlab.com/tzip/tzip/-/merge_requests/67) is about updating transfer parameter which is done already here and mostly 
for their inner code conversion refactoring and document restructure, so only [!66](https://gitlab.com/tzip/tzip/-/merge_requests/66) is left to be
done.

## Related issue(s)

Resolves #68 

#### Related changes (conditional)

- Tests
  - [X] If I added new functionality, I added tests covering it.
  - [X] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [X] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock
  - [X] I updated [changelog](../tree/master/ChangeLog.md) unless I am sure my changes are
        not essential.

#### Stylistic guide (mandatory)

- [X] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [X] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by gromakovsky** [comment]
AFAIU, currently it only accounts for !67, right? I see that !67 was closed and `total_supply` was removed in !66. So most likely it will be removed eventually.

Apart from that note that !66 was is not merged yet, so maybe some changes will happen there, but most likely there will be no substantial changes, so I think you can already start incorporating !66 changes, but we won't merge them before they land to tzip's master.

**Comment by awkure** [comment]
Oh, I mean that we have already done !67 earlier and this MR accommodates changes from !66 (which is actually closed but not merged) so I believe that this issue may be closed as well, but it needs to be re-checked.

**Comment by awkure** [comment]
Ok, `token_metadata` type seems to be updated and an additional entrypoint `Token_metadata_registry` was added. I wish they at least explain what they have done somewhere before merging these changes to master.

**Comment by gromakovsky** [comment]
Actually there are quite a lot of changes in !66. And it was merged recently btw.

File: docs/specification.md [hunk 1]
```diff
@@ -324,7 +325,7 @@ token_metadata_response =
 
 token_metadata =
```
Comment (sras) [inline docs/specification.md:326]: Here and above `token_metadata` type alias is used. I think one of them should be `token_metadata_request`.

Comment (sras) [reply docs/specification.md:326]: Also, do we need this entrypoint? Because aren't we implementing this alternative? 

>Contract storage MUST have a big_map that maps token_id -> token_metadata
and annotated %token_metadata

File: haskell/test/Lorentz/Contracts/Test/FA2.hs [hunk 2]
```diff
@@ -844,7 +786,7 @@ fa2Spec fa2Originate = do
 
           withSender commonOperator $ lCallEP fa2contract (Call @"Transfer") transfers
 
-          validate $ Left fa2SenderHookUndefined
+          validate $ Left fa2SenderHookFailed
```
Comment (sras) [inline haskell/test/Lorentz/Contracts/Test/FA2.hs:789]: Why was undefined hook error changed to hook failed error?

File: docs/specification.md [hunk 3]
```diff
@@ -129,6 +129,7 @@ Full list:
 * [`balance_of`](#balance_of)
 * [`total_supply`](#total_supply)
```
Comment (sras) [inline docs/specification.md:130]: Shouldn't we remove `total_supply` from here and the corresponding section below as well?

File: docs/specification.md [hunk 4]
```diff
@@ -324,7 +325,7 @@ token_metadata_response =
 
 token_metadata =
   ( list token_id                           :token_ids
-  , contract (list token_metadata_response) :callback
+  , contract (list token_metadata) :callback
```
Comment (sras) [inline docs/specification.md:328]: Shouldn't this be a lambda instead?

File: docs/specification.md [hunk 5]
```diff
@@ -766,6 +783,8 @@ Note that operator relation is not transitive.
 
 One more uncertainty is related to the `token_metadata` entrypoint: whether metadata is constant or can be changed during contract's lifetime.
 We assume that token metadata stays constant forever.
+Upd: We have an additional entrypoint that holds all token metadata
```
Comment (sras) [inline docs/specification.md:786]: >Upd:

Not sure if this is a left over.

File: haskell/test-common/Lorentz/Contracts/Test/Common.hs [hunk 6]
```diff
@@ -211,6 +212,14 @@ constructSingleTransfer
 constructSingleTransfer from to amount
     = [(from, (#txs .! [(to, (#token_id .! 0, amount))]))]
 
+-- | Token metadata parameter that takes list of token ids and
+-- assumes that each of which resulting metadata is always valid
+assertAnyTokenMetaData :: [Natural] -> FA2.TokenMetaDataParam
+assertAnyTokenMetaData ids =
```
Comment (sras) [inline haskell/test-common/Lorentz/Contracts/Test/Common.hs:218]: I am not sure if it is proper to call this an assertion, because it can be something else, like posting token metadata to a different contract. right?

File: docs/specification.md (hunk 5 above)
Comment (awkure) [reply docs/specification.md:786]: No, it's an commentary that  needs to be discussed

File: haskell/test/Lorentz/Contracts/Test/FA2.hs (hunk 2 above)
Comment (awkure) [reply haskell/test/Lorentz/Contracts/Test/FA2.hs:789]: To match `FA2_RECEIVER_HOOK_FAILED` error as is defined in FA2 spec 🤷

File: docs/specification.md (hunk 1 above)
Comment (awkure) [reply docs/specification.md:326]: Well, If it's required by FA2 so I believe that we should leave it as is here as well

File: docs/specification.md (hunk 3 above)
Comment (awkure) [reply docs/specification.md:130]: yep

File: haskell/test/Lorentz/Contracts/Test/FA2.hs (hunk 2 above)
Comment (sras) [reply haskell/test/Lorentz/Contracts/Test/FA2.hs:789]: But here the hook is undefined here on the reciever, so shouldn't it be the undefined error as per FA2?

File: docs/specification.md (hunk 1 above)
Comment (sras) [reply docs/specification.md:288]: But AFAIU FA2 specifies

>Token registry contract MUST implement one of two ways to expose token metadata
for off-chain clients:

So it can be either one of those. I ask because @gromakovsky suggested here [1] that we can use the simpler alternative of just leaving the metadata in the storage, without actually implementing the entrypoint alternative.

[1] https://serokell.slack.com/archives/CUMAQ3PL4/p1592391873179000

Comment (awkure) [reply docs/specification.md]: Alright, have removed this entrypoint completely

File: haskell/test/Lorentz/Contracts/Test/FA2.hs (hunk 2 above)
Comment (awkure) [reply haskell/test/Lorentz/Contracts/Test/FA2.hs]: Ah, you're right, renamed back

File: haskell/test/SMT.hs [hunk 7]
```diff
@@ -7,29 +7,30 @@ module SMT
   ) where
 
 import qualified Data.Map as Map
+import Data.Maybe (fromJust)
```
Comment (gromakovsky) [inline haskell/test/SMT.hs:10]: Is it used here?

File: ligo/stablecoin/spec.ligo [hunk 8]
```diff
@@ -79,8 +79,7 @@ end
 type balance_of_params is michelson_pair_right_comb(balance_of_params_)
 
 type token_metadata_ is record
-  token_id  : token_id
```
Comment (gromakovsky) [inline ligo/stablecoin/spec.ligo:96]: Why did you remove `token_id` from here? AFAIU it contradicts FA2. We even have discussed it with Sandeep in [this thread](https://serokell.slack.com/archives/CUMAQ3PL4/p1592391873179000) before.

File: haskell/test/SMT.hs (hunk 7 above)
Comment (gromakovsky) [reply haskell/test/SMT.hs:10]: @awkure ping

File: haskell/test/SMT.hs [hunk 9]
```diff
@@ -7,29 +7,30 @@ module SMT
...
 import Data.Typeable (cast)
 import Fmt
```
Comment (gromakovsky) [inline haskell/test/SMT.hs:12]: This import should be explicitl

**Comment by gromakovsky** [comment]
You can cleanup existing commits btw, I have checked them (now there are some commits that fix previous ones and should be squashed).

File: docs/specification.md [hunk 10]
```diff
@@ -271,90 +270,21 @@ Parameter (in Michelson):
...
-```
-
-- This entrypoint MUST follow the FA2 requirements.
-
-- Since the contract supports only a single token type, all `token_id` values MUST be 0.
-  They are passed because FA2 requires that.
+- Return contract address that holds token metadata.
 
-- Token metadata for 0 `token_id` is constant and should be hardcoded in contract code or put into storage during origination.
+- Since the contract owns its token metadata the returned value of this entrypoint call will always be equal to `SELF`.
```
Comment (gromakovsky) [inline docs/specification.md:287]: I guess we should write somewhere that this contract (i.e. our stablecoin contract) must have a `big_map` with a certain annotation and type.

**Comment by gromakovsky** [comment]
```
Problem: With recent FA2 changes `token_metadata` was restructured
so that its parameter accepts `handler` lambda rather than
a callback contract. Also `token_metadata_registry` entryopint
was added that returns an address of a contract that stores actual
token metadata.

Solution: Add `token_metadata_registry` entrypoint, restructure
`token_metadata` parameter and update specification and hasskell
types with tests as well.
```

1. ↑ It's a bit outdated now.
2. There are some unresolved discussions above.
3. Minor: the last commit has some changes that fix previous commits:
  + Restoration of some changed errors (amends the first commit)
  + Removal of `total_supply` from the specification.

**Comment by gromakovsky** [comment]
Also it needs to be rebased.

**Comment by gromakovsky** [comment]
Alright, waiting for rebase so that I can take a look at the contract from CI.
